packages/network/lib/agent.ts

```diff
@@ -293,7 +371,9 @@ class HttpsAgent extends https.Agent {
       options.port = 443
     }
 
-    super.addRequest(req, options)
+    mergeCAOption(options).then((options) => {
+      super.addRequest(req, options)
+    })
```

Comment:
Originally, we decided to make the agent use callback patterns throughout because callbacks are synchronous, whereas `.then` is asynchronous. So effectively this change adds 1 tick to all outbound HTTPS requests, which can result in a cost of >1ms per request before it's sent outbound, depending on how many tasks are queued for the current or next tick.

What do you think about changing this to still issue requests synchronously where possible? Something like this:

```ts
/** top of file, or on the first HTTPS request **/
let caOptions
const caOptionsPromise = getCaOptions().then(options => {
  // with no extra CA, caOptions is { ca: tls.rootCertificates } or {}
  caOptions = options
  return caOptions
})

/** in addRequest **/
if (caOptions) {
  return super.addRequest(req, merge(options, caOptions))
}

caOptionsPromise.then(caOptions => {
  super.addRequest(req, merge(options, caOptions))
})
```

Reply:
It feels sorta like a micro-optimization at first, because so many Promises are chained to get to this point, but I think it's worth doing because we're just using the Promise to cache the resolved value here. That can be done synchronously and HTTPS requests can get sent out slightly faster.

Reply:
I like this idea. Implemented it and it's good to go on the PR now.